orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java

```diff
@@ -0,0 +1,231 @@
+      return (float) ((DoubleColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class DoubleReader implements OrcValReader<Double> {
+    private static final DoubleReader INSTANCE = new DoubleReader();
+
+    private DoubleReader() {
+    }
+
+    @Override
+    public Double nonNullRead(ColumnVector vector, int row) {
+      return ((DoubleColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class ByteReader implements OrcValReader<Byte> {
+    private static final ByteReader INSTANCE = new ByteReader();
+
+    private ByteReader() {
+    }
+
+    @Override
+    public Byte nonNullRead(ColumnVector vector, int row) {
+      return (byte) ((LongColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class BytesReader implements OrcValReader<byte[]> {
+    private static final BytesReader INSTANCE = new BytesReader();
+
+    private BytesReader() {
+    }
+
+    @Override
+    public byte[] nonNullRead(ColumnVector vector, int row) {
+      BytesColumnVector bytesVector = (BytesColumnVector) vector;
+
+      return Arrays.copyOfRange(
+          bytesVector.vector[row], bytesVector.start[row], bytesVector.start[row] + bytesVector.length[row]);
+    }
+  }
+
+  public abstract static class StructReader<T> implements OrcValReader<T> {
+    private final OrcValReader<?>[] readers;
+    private final int[] positions;
+    private final Object[] constants;
+
+    protected StructReader(List<OrcValReader<?>> readers) {
+      this.readers = readers.toArray(new OrcValReader[0]);
+      this.positions = new int[0];
+      this.constants = new Object[0];
+    }
+
+    protected StructReader(List<OrcValReader<?>> readers, Types.StructType struct, Map<Integer, ?> idToConstant) {
+      this.readers = readers.toArray(new OrcValReader[0]);
+      List<Types.NestedField> fields = struct.fields();
+      List<Integer> positionList = Lists.newArrayListWithCapacity(fields.size());
+      List<Object> constantList = Lists.newArrayListWithCapacity(fields.size());
+      for (int pos = 0; pos < fields.size(); pos += 1) {
+        Types.NestedField field = fields.get(pos);
+        Object constant = idToConstant.get(field.fieldId());
+        if (constant != null) {
+          positionList.add(pos);
+          constantList.add(idToConstant.get(field.fieldId()));
+        }
+      }
+
+      this.positions = positionList.stream().mapToInt(Integer::intValue).toArray();
+      this.constants = constantList.toArray();
+    }
+
+    protected abstract T create();
+
+    protected abstract T reuseOrCreate();
+
+    protected abstract void set(T struct, int pos, Object value);
+
+    public OrcValReader<?> reader(int pos) {
+      return readers[pos];
+    }
+
+    @Override
+    public T nonNullRead(ColumnVector vector, int row) {
+      StructColumnVector structVector = (StructColumnVector) vector;
+      return readInternal(create(), structVector.fields, row);
+    }
+
+    public T read(VectorizedRowBatch batch, int row) {
```

Comment:
Instead of having an interface that uses `VectorizedRowBatch`, why not just have the top-level reader create a `StructColumnVector` and pass the batch's columns using that?

You can even make a reusable `StructColumnVector` to avoid needing to create a new one for every batch if you want:

```java
  class ReusableStructColumnVector extends StructColumnVector {
    public ReusableStructColumnVector(int len, ColumnVector... fields) {
      super(len, fields);
    }

    public ReusableStructColumnVector replaceVectors(ColumnVector[] fields) {
      this.fields = fields;
      return this;
    }
  }
```

Then we'd only need one `OrcValueReader` interface.

Reply:
I think I did what you are suggesting before, but ran into issues. Let me try again to get the actual erorr.

Reply:
It worked this time.